src/olympia/shelves/forms.py

```diff
@@ -15,14 +15,19 @@ class Meta:
                   'footer_text', 'footer_pathname',)
 
     def clean(self):
-        data = self.cleaned_data
-        criteria = data['criteria']
+        cleaned_data = super(ShelfForm, self).clean()
+
+        if any(self.errors):
+            return self.errors
+
+        criteria = cleaned_data.get('criteria')
+        endpoint = cleaned_data.get('endpoint')
```

Comment:
these couple of lines that replace `data['endpoint']` with `data.get('endpoint')` (and same with `criteria`) are the fix for the issue, aren't they?  I don't understand what the rest of the change in the patch is for.

Reply:
There would still be a `KeyError` if any of the required fields were left blank when attempting to save the form. The fix of the issue was to return any errors first with:
`if any(self.errors):`
 `            return self.errors`

I guess I didn't necessarily have to replace `data['endpoint']` with `data.get('endpoint')`, but I read that was a better way to write it. The rest of the changes were just to move the error message above the field, as opposed to the top of the form, which is more of a preference and not necessary for the the patch. Please let me know if you want me to revert those revisions.